sbom-utility-scripts/scripts/merge-cachi2-sboms-script/merge_cachi2_sboms.py

```diff
@@ -148,6 +299,199 @@ def _merge_tools_metadata(syft_sbom: dict[Any, Any], cachi2_sbom: dict[Any, Any]
+        and value is spdx element.
+        """
+        relations_map = {}
+        relations_inverse_map = {}
+
+        for relation in relationships:
+            relations_map.setdefault(relation["spdxElementId"], []).append(relation["relatedSpdxElement"])
+            relations_inverse_map[relation["relatedSpdxElement"]] = relation["spdxElementId"]
+
+        for parent_element in relations_map.keys():
+            if parent_element not in relations_inverse_map:
+                break
+        return parent_element, relations_map, relations_inverse_map
+
+    def calculate_root_package(root_element, map, inverse_map):
+        """Calculate root package from relationship map.
+        Root package is considered as package which contains other packages and
+        is described by the document itself.
+        """
+        root_package = None
+        for r, contains in map.items():
+            if contains and inverse_map.get(r) == root_element:
+                root_package = r
+        return root_package
+
+    relationships = []
+
+    root_element1, map1, inverse_map1 = map_relationships(relationships1)
+    root_element2, map2, inverse_map2 = map_relationships(relationships2)
+    package_ids = [package["SPDXID"] for package in packages]
+
+    root_package1 = calculate_root_package(root_element1, map1, inverse_map1)
+    root_package2 = calculate_root_package(root_element2, map2, inverse_map2)
+
+    for relation in relationships2:
+        _relation = relation.copy()
+
+        # If relations is Root decribes middle element, skip it
+        if (
+            _relation["relatedSpdxElement"] == root_package2
+            and _relation["spdxElementId"] == root_element2
+            and _relation["relationshipType"] == "DESCRIBES"
+        ):
+            continue
+        # if spdxElementId is root_element2, replace it with root_element1
+        # if not and relatedSpdxElement is root_element2, replace it with root_element1
+        if _relation["spdxElementId"] == root_element2:
+            _relation["spdxElementId"] = root_element1
+        elif relation["relatedSpdxElement"] == root_element2:
+            _relation["relatedSpdxElement"] = root_element1
+        if _relation["spdxElementId"] == root_package2:
+            _relation["spdxElementId"] = root_package1
+        if _relation["relatedSpdxElement"] == root_package2:
+            _relation["relatedSpdxElement"] = root_package1
+
+        # include only relations to packages which exists in merged packages.
+        if _relation["relatedSpdxElement"] in package_ids:
+            relationships.append(_relation)
+        elif _relation["spdxElementId"] in package_ids:
+            relationships.append(_relation)
+
+    for relation in relationships1:
+        _relation = relation.copy()
+        if relation["relatedSpdxElement"] in package_ids:
+            relationships.append(_relation)
```

Comment:
Why do we only look at `relatedSpdxElement` for document 1 when we look at both `spdxElementId` and `relatedSpdxElement` for document 2?

Reply:
Because in the first doc spdxElementId can be root element and in that case even relationships pointing to removed packages would be included